hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -628,8 +627,8 @@ private void authenticate(final Address target, final ClientConnection connectio
         final ClientPrincipal principal = getPrincipal();
         ClientMessage clientMessage = encodeAuthenticationRequest(asOwner, client.getSerializationService(), principal);
         ClientInvocation clientInvocation = new ClientInvocation(client, clientMessage, null, connection);
+        clientInvocation.setHardTimeoutMillis(connectionTimeoutMillis);
```

Comment:
why is this extra setting needed while we already have connectionTimeoutMillis?

Reply:
`connectionTimeoutMillis` was checked via countdownLatch on AuthCallback . We changed it to be checked via a scheduled task in https://github.com/hazelcast/hazelcast/pull/12163. 

First one had thread race problems. 
The second one is causing threads to be closed late in case of a shutdown. 

I have introduced `clientInvocation.setHardTimeoutMillis`  to solve this. I did not introduce a new config, I am just checking `connectionTimeoutMillis` in a different way so that it does not have any race problems and does not cause a delay on shutdown.